src/components/currency-converter.tsx

```diff
@@ -0,0 +1,243 @@
+    const limit = currencyLimits[fromCurrency];
+    if (amount === null || rate === null || amount.isZero()) {
+      setError("Amount cannot be empty or zero");
+      return;
+    }
+    if (limit && amount.isGreaterThan(limit)) {
+      setError(`Amount exceeds limit for ${fromCurrency}: ${limit.toString()}`);
+      return;
+    }
+    const newAmount = amount.multipliedBy(rate);
+    setConvertedAmount(newAmount);
+    setIsInitialConversion(true);
+  };
+
+  const handleToAmountChange = (value: BigNumber | null) => {
+    setConvertedAmount(value);
+    setLastEditedField("to");
+  };
+
+  useEffect(() => {
+    const limit = currencyLimits[fromCurrency];
+    if (amount === null) return;
+
+    if (limit && amount.isGreaterThan(limit)) {
+      setError(`Amount exceeds limit for ${fromCurrency}: ${limit.toString()}`);
+    } else {
+      setError(null);
+    }
+  }, [amount, convertedAmount, fromCurrency]);
+
+  useEffect(() => {
+    const limit = currencyLimits[fromCurrency];
+    if (!exchangeRate) return;
+
+    if (limit && amount !== null && amount.isGreaterThan(limit)) {
+      setError(`Amount exceeds limit for ${fromCurrency}: ${limit.toString()}`);
+      return;
+    }
+
+    if (lastEditedField === "from" && amount !== null) {
+      const newConvertedAmount = amount
+        .multipliedBy(exchangeRate)
+        .decimalPlaces(2);
+      if (
+        convertedAmount === null ||
+        !newConvertedAmount.isEqualTo(convertedAmount)
+      ) {
+        setConvertedAmount(newConvertedAmount);
+      }
+    } else if (lastEditedField === "to" && convertedAmount !== null) {
+      const newAmount = convertedAmount
+        .dividedBy(exchangeRate)
+        .decimalPlaces(2);
+      if (limit && newAmount.isGreaterThan(limit)) {
+        setError(
+          `Amount exceeds limit for ${fromCurrency}: ${limit.toString()}`,
+        );
+        console.log(error);
+        return;
+      } else if (amount === null || !newAmount.isEqualTo(amount)) {
+        setAmount(newAmount);
+      }
+    }
+  }, [amount, convertedAmount, lastEditedField, exchangeRate]);
+
+  useEffect(() => {
+    if (fromCurrency && toCurrency && amount) {
+      fetchExchangeRate(fromCurrency, toCurrency).then((rate) => {
+        if (rate !== null) {
+          const newConvertedAmount = amount.multipliedBy(rate).decimalPlaces(2);
+          setConvertedAmount(newConvertedAmount);
+        }
+      });
+    }
+  }, [fromCurrency, toCurrency]);
+
+  const handleFromAmountChange = (value: BigNumber | null) => {
+    if (value !== null) {
+      const limit = currencyLimits[fromCurrency];
+      if (limit && value.isGreaterThan(limit)) {
+        setError(
+          `Amount exceeds limit for ${fromCurrency}: ${limit.toString()}`,
+        );
+      } else {
+        setError(null);
+      }
+    }
+    setAmount(value);
+    setLastEditedField("from");
+  };
+
+  return (
+    <div className="max-w-xl mx-auto my-10 p-12 bg-white rounded-lg shadow-xl">
```

Comment:
this whole calc could be treated as a `form`, because it consists of two inputs and a submit button, so it's a `form` :)